### PR TITLE
Add API for job definitions

### PR DESCRIPTION
**Pull request by t8m8**
I added API for job definitions.

You can use this as follow.

```
# index
$ curl -s -u user:password http://example.com/v1/definitions

# create
$ curl -s -u user:password -XPOST http://example.com/v1/definitions \
  -d "name=job_name" \
  -d "description=your_description" \
  -d "script=noop:" \
  -d "notify_cancellation=1" \
  -d "hipchat_room=" \
  -d "hipchat_notify_finished=1"\
  -d "suspended=false" \
  -d "prevent_multi=1" \
  -d "hipchat_additional_text=" \
  -d "text_tags=" \
  -d "api_allowed=1" \
  -d "slack_channel=" \
  -d "webhook_url=" \
  -d "user_id[]=1"
 
# show
$ curl -s -u user:password http://example.com/v1/definitions/1

# update
$ curl -s -u user:password -XPUT http://example.com/v1/definitions/1 \
  -d "name=job_name" \
  -d "description=your_description" \
  -d "script=noop:" \
  -d "notify_cancellation=1" \
  -d "hipchat_room=" \
  -d "hipchat_notify_finished=1"\
  -d "suspended=false" \
  -d "prevent_multi=1" \
  -d "hipchat_additional_text=" \
  -d "text_tags=" \
  -d "api_allowed=1" \
  -d "slack_channel=" \
  -d "webhook_url="

# delete
$ curl -s -u user:password -XDELETE http://example.com/v1/definitions/1
```

### CODE REVIEW

**Comment by eisuke** [comment]
@cookpad/dev-infra

File: app/controllers/kuroko2/api/job_definitions_controller.rb [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+
+    if definition.save
+      user_ids.each do |user_id|
+        user = Kuroko2::User.active.find(user_id)
+        user.stars.create(job_definition: definition) if user.google_account?
+      end
+      
+      @resource = Kuroko2::Api::JobDefinitionResource.new(definition)
+    else
+      raise HTTP::Forbidden.new("#{definition.name}: #{definition.errors.full_messages.join()}")
```
Comment (eisuke) [inline app/controllers/kuroko2/api/job_definitions_controller.rb]: The `Forbidden` response seems not appropriate.

File: app/controllers/kuroko2/api/job_definitions_controller.rb [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+    definition = Kuroko2::JobDefinition.find(params[:id])
+    @resource = Kuroko2::Api::JobDefinitionResource.new(definition)
+  end
+  
+  def update_resource
+    definition = Kuroko2::JobDefinition.find(params[:id])
+    unless definition.api_allowed?
+      raise_not_allowed(definition)
+    end
+    definition.update_attributes(definition_params(params))
```
Comment (eisuke) [inline app/controllers/kuroko2/api/job_definitions_controller.rb]: Use `#update!`, or do validation.
How about using `weak_parameter` gem ? https://github.com/r7kamura/weak_parameters

File: app/controllers/kuroko2/api/job_definitions_controller.rb [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+    @resource = Kuroko2::Api::JobDefinitionResource.new(definition)
+  end
+
+  def destroy_resource
```
Comment (eisuke) [inline app/controllers/kuroko2/api/job_definitions_controller.rb]: I feel `destroy` is not necessary. What is the use case of this endpoint?

File: app/controllers/kuroko2/api/job_definitions_controller.rb [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+    definition = Kuroko2::JobDefinition.find(params[:id])
+    unless definition.api_allowed?
```
Comment (eisuke) [inline app/controllers/kuroko2/api/job_definitions_controller.rb]: ditto https://github.com/cookpad/kuroko2/pull/87/files#r135692164

File: app/controllers/kuroko2/api/job_definitions_controller.rb [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+  private
+
+  def require_resources
+    definitions = Kuroko2::JobDefinition.all
+    @resources = definitions.map {|definition| Kuroko2::Api::JobDefinitionResource.new(definition) }
+  end
+
+  def create_resource
+    definition = Kuroko2::JobDefinition.new(definition_params(params))
+    unless definition.api_allowed?
```
Comment (eisuke) [inline app/controllers/kuroko2/api/job_definitions_controller.rb]: ditto https://github.com/cookpad/kuroko2/pull/87/files#r135692164

File: spec/requests/api/job_definitions_spec.rb [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+          'name' => definition.name,
+          'description' => definition.description,
+          'script' => definition.script
+        }]
+      )
+      expect(response.status).to eq(200)
+    end
+  end
+
+  describe 'POST /v1/definitions' do
```
Comment (eisuke) [inline spec/requests/api/job_definitions_spec.rb]: I want to tests with invalid parameters.

File: spec/requests/api/job_definitions_spec.rb [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+          'name' => definition.name,
+          'description' => definition.description,
+          'script' => definition.script
+        }
+      )
+      expect(response.status).to eq(200)
+    end
+  end
+
+  describe 'PUT /v1/definitions/:id' do
```
Comment (eisuke) [inline spec/requests/api/job_definitions_spec.rb]: ditto https://github.com/cookpad/kuroko2/pull/87/files#r135695226

File: spec/requests/api/job_definitions_spec.rb [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+        Kuroko2::JobDefinition.count
+      }.by(1)
+      expect(result['name']).to eq(params[:name])
+      expect(result['description']).to eq(params[:description])
+      expect(result['script']).to eq(params[:script])
+      expect(response.status).to eq(201)
+    end
+  end
+
+  describe 'GET /v1/definitions/:id' do
```
Comment (eisuke) [inline spec/requests/api/job_definitions_spec.rb]: I want to test if `job_definition` not found.

File: app/controllers/kuroko2/api/job_definitions_controller.rb [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+  end
+
+  def require_resource
+    definition = Kuroko2::JobDefinition.find(params[:id])
+    @resource = Kuroko2::Api::JobDefinitionResource.new(definition)
+  end
+  
+  def update_resource
+    definition = Kuroko2::JobDefinition.find(params[:id])
+    unless definition.api_allowed?
```
[the remark this answers is not in the document]
Comment (t8m8) [reply app/controllers/kuroko2/api/job_definitions_controller.rb]: I have fixed it in 55b4da1.

File: app/controllers/kuroko2/api/job_definitions_controller.rb (hunk 2 above)
Comment (t8m8) [reply app/controllers/kuroko2/api/job_definitions_controller.rb]: I have fixed it in 55b4da1.

File: app/controllers/kuroko2/api/job_definitions_controller.rb (hunk 4 above)
Comment (t8m8) [reply app/controllers/kuroko2/api/job_definitions_controller.rb]: I have fixed it in 55b4da1.

File: app/controllers/kuroko2/api/job_definitions_controller.rb (hunk 5 above)
Comment (t8m8) [reply app/controllers/kuroko2/api/job_definitions_controller.rb]: I have fixed it in 55b4da1.

File: app/controllers/kuroko2/api/job_definitions_controller.rb [hunk 10]
```diff
@@ -0,0 +1,81 @@
+class Kuroko2::Api::JobDefinitionsController < Kuroko2::Api::ApplicationController
+  include Garage::RestfulActions
+
+  private
+
+  def require_resources
```
[the remark this answers is not in the document]
Comment (t8m8) [reply app/controllers/kuroko2/api/job_definitions_controller.rb]: I have removed it in 11be3c8.

File: app/controllers/kuroko2/api/job_definitions_controller.rb (hunk 3 above)
Comment (t8m8) [reply app/controllers/kuroko2/api/job_definitions_controller.rb]: I have removed it in 73543d4.

File: spec/requests/api/job_definitions_spec.rb (hunk 8 above)
Comment (t8m8) [reply spec/requests/api/job_definitions_spec.rb]: I have fixed it in 939b69d

File: spec/requests/api/job_definitions_spec.rb (hunk 6 above)
Comment (t8m8) [reply spec/requests/api/job_definitions_spec.rb]: I have fixed it in 939b69d.

File: app/controllers/kuroko2/api/job_definitions_controller.rb (hunk 1 above)
Comment (t8m8) [reply app/controllers/kuroko2/api/job_definitions_controller.rb]: I have fixed it in 939b69d.

File: app/controllers/kuroko2/api/job_definitions_controller.rb [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+    definition = Kuroko2::JobDefinition.new(definition_params(params))
+    unless definition.api_allowed?
+      raise_not_allowed(definition)
+    end
+    user_ids = admin_id_params(params)
+    definition.admins = Kuroko2::User.active.with(user_ids)
+
+    if definition.save
+      user_ids.each do |user_id|
+        user = Kuroko2::User.active.find(user_id)
```
[the remark this answers is not in the document]
Comment (t8m8) [reply app/controllers/kuroko2/api/job_definitions_controller.rb]: I have fixed it in 83c1423.

File: spec/requests/api/job_definitions_spec.rb (hunk 7 above)
Comment (t8m8) [reply spec/requests/api/job_definitions_spec.rb]: I have fixed it in c360037.

File: spec/requests/api/job_definitions_spec.rb [hunk 12]
```diff
@@ -0,0 +1,179 @@
...
+          api_allowed: 1,
+          slack_channel: "",
+          webhook_url: "",
+        }
+      end
+
+      it 'updates a definition' do
+        expect {
+          put "/v1/definitions/#{definition.id}", params: params, env: env
+        }.to raise_error
```
Comment (eisuke) [inline spec/requests/api/job_definitions_spec.rb]: Please check reponse status code.

File: app/controllers/kuroko2/api/job_definitions_controller.rb (hunk 2 above)
Comment (t8m8) [reply app/controllers/kuroko2/api/job_definitions_controller.rb]: I have modified it in 4982de6.
I used #update!.

File: app/controllers/kuroko2/api/job_definitions_controller.rb [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+    definition.admins = Kuroko2::User.active.with(user_ids)
+
+    if definition.save
+      definition.admins.each do |user|
+        user.stars.create(job_definition: definition) if user.google_account?
+      end
+      
+      @resource = Kuroko2::Api::JobDefinitionResource.new(definition)
+    else
+      raise WeakParameters::ValidationError.new("#{definition.name}: #{definition.errors.full_messages.join()}")
```
Comment (eisuke) [inline app/controllers/kuroko2/api/job_definitions_controller.rb]: It has nothing to do with `weak_parameters`.
Please use `Kuroko2::Api::ApplicationController#respond_with_error`.

Comment (t8m8) [reply app/controllers/kuroko2/api/job_definitions_controller.rb]: I have fixed it in f81b2ce.

File: spec/requests/api/job_definitions_spec.rb (hunk 12 above)
Comment (t8m8) [reply spec/requests/api/job_definitions_spec.rb]: I have fixed it in f81b2ce.

**Pull request merged**